add new flag from Xcode 9

diederich opened the pull request:
There's a new Swift compilation flag that Xcode 9.3 wants to add. No documentation for it yet though :-/

fabb commented:
here's the documentation: https://swift.org/blog/osize/

diederich commented:
Oh - I haven't checked this against Xcode 9.2...

fabb commented:
i guess `SWIFT_OPTIMIZATION_LEVEL` should be set to `-O` once Xcode 9.3 lands

diederich commented:
Oh, interesting, that does surprise me a little TBH. I this also for incremental builds? - I'd imagine it needs to redo the whole lot instead of single files? 
(Honestly I haven't  really looked into swift compilation times at all - so I'm happy to add it if it speeds up things)

fabb commented:
Yeah, wholemodule + no optimization is THE performance boost for debug build times:
https://jobs.zalando.com/tech/blog/improving-swift-compilation-times-from-12-to-2-minutes/

From the swift blog:
> After performing whole-module optimizations in the SIL optimizer the module is split again into multiple parts. The LLVM backend processes the split parts in multiple threads. It also avoids re-processing of a part if that part didn’t change since the previous build.
https://swift.org/blog/whole-module-optimizations/

diederich commented:
Thanks for the links! Super interesting. 
I'm not sold on the build times though. I'm not sure we can read out of those articles that incremental build times are better with "wholemodule" turned on. 

The Zalando blog post compares "clean build times" for single-file changes: 
> Changing a file that most of the app depends on now takes only 1 minute instead of the full recompilation time. 

And the swift blog says:

> It also avoids re-processing of a part if that part didn’t change since the previous build.

Which is great, but I'd assume this mainly means everything before the SIL stage. Once all the files got "munched" together, it'll be hard to do. And in the end they finish with:

>  And you get this high performance with much better compile times than typical to monolithic whole-program optimization approaches.

Which is awesome, but also kind of means "if you want it to be as fast as possible, turn it off" :-) 

--- 

I did some quick super minimal testing with a project with only 26 swift files (and lots of objective-c):
https://gist.github.com/diederich/bb55910311ac432ddbcb2c212d8ec4d6
Clean builds are indeed faster, where as incremental builds are in that case 50% slower (I'm sure this is not representative, but I think makes somewhat sense)

I just checked with Xcode 9.3 and for new projects the compilation mode is still set to 'Single File'. Should we just go with the default for now?

fabb commented:
Ok from my side, as we‘ll override the setting in our project either way, as for us it makes a huge difference.

fabb commented:
ah ok, i thought with "Should we just go with the default for now?" you meant that you would set `SWIFT_COMPILATION_MODE = singlefile` explicitly in `Debug.xcconfig`

diederich commented:
Oh - uhm, sorry I did miss that. But should we? Or just leave as default?

fabb commented:
🤷‍♂️ for my projects i like to make settings explicit rather depending on xcode defaults

diederich commented:
I'm happy to leave as is. We even get it updated for free if Xcode eventually switches it over. Feel free to add it though.
